### ISSUE TITLE
Missing SKAT/SKATO/ACATV/ACATO For Certain Masks

### ISSUE THREAD

**Issue by christopherdkoch** [rgcgithub/regenie#275]
Thanks for a great tool and v3 updates! I'm trying to run variance component tests for a series of masks and genes. For some gene/mask combinations for which I get burden outputs I'm not getting SKAT*/ACAT* outputs. Thoughts on what could be causing this?

**Comment by joellembatchou** [comment]
Hi,

Can you rerun Regenie for a mask where you don't have SKAT/ACAT results using `--debug`?

**Comment by christopherdkoch** [comment]
Hi @joellembatchou, I reran the tool with a single gene and single mask. Burden and ADD-BURDEN-ACAT worked, but not SKAT/ACAT.

Below is what I have in the .e file. I can send the standard out as well, but need to remove some file paths/names.

Q_SKAT for all masks:
0 0
Q_BURDEN for all masks:
0 0

**Comment by christopherdkoch** [comment]
This is all the gets printed in the standard error output. Yes, I ran ACATV also but did not get outputs for that either for this mask/gene. This is a QT phenotype. I can check if the problem exists for BTs.

`Start time: Tue Apr  5 22:12:39 2022

              |=============================|
              |      REGENIE v3.0.1.gz      |
              |=============================|

Copyright (c) 2020-2022 Joelle Mbatchou, Andrey Ziyatdinov and Jonathan Marchini.
Distributed under the MIT License.
Compiled with Boost Iostream library.
Using Intel MKL with Eigen.

Log of output saved in file : [FILENAME]

Options in effect:
  --step 2 \
  --threads 1 \
  --bed [FILENAME] \
  --covarFile [FILENAME] \
  --phenoFile [FILENAME] \
  --phenoCol [PHENO] \
  --bsize 100 \
  --pred [FILENAME] \
  --out [FILENAME] \
  --aaf-bins 0.05 \
  --anno-file [FILENAME] \
  --set-list [FILENAME] \
  --mask-def [FILENAME] \
  --write-mask \
  --apply-rint \
  --minMAC 1 \
  --vc-tests skat,skato,acatv,acato \
  --vc-maxAAF 0.05 \
  --joint acat \
  --debug \
  --range 8:19759228-19824769

Association testing mode (joint tests) with fast multithreading using OpenMP
 * bim              : [[FILENAME]] n_snps = 4982410
   -number of variants after filtering on range = 300
 * fam              : [[FILENAME]] n_samples = 30108
 * bed              : [[FILENAME]]
 * phenotypes       : [[FILENAME]] n_pheno = 1
   -dropping observations with missing values at any of the phenotypes
   -number of phenotyped individuals with no missing data = 27519
 * covariates       : [[FILENAME]] n_cov = 14
   -number of individuals with covariate data = 30106
 * number of individuals used in analysis = 27519
   -applying RINT to all phenotypes
 * LOCO predictions : [[FILENAME]]
   -file [[FILENAME]] for phenotype '[PHENO]'
   -residualizing and scaling phenotypes...done (0ms) 
 * annotations      : [[FILENAME]] 
   +number of annotations categories = 7
 * masks            : [[FILENAME]] n_masks = 1
 * aaf cutoffs      : [ 1 : 0.05 ] + singletons
 * set file         : [[FILENAME]] n_sets = 1
WARNING: Detected 19432 sets with only unknown variants (these are ignored).
 * # threads        : [1]
 * # tested sets    : [1]
 * max block size   : [100]
 * rule used to build masks : max
 * computing gene-based tests for each set of variants included in a mask (rho=[0,0.01,0.04,0.09,0.16,0.25,0.5,1])
  -variants with MAC <= 10 are collapsed into a mask
  -weights are obtained from Beta(MAF,1,25)
 * approximate memory usage : 28MB
 * user specified to write masks (in PLINK bed format)
 * using minimum MAC of 1 (masks with lower MAC are ignored)

Chromosome 8 [1 sets in total]
   -reading loco predictions for the chromosome...done (19ms) 
 set [1/1] : 9 variants...1 chunks
     -reading in genotypes, computing gene-based tests and building masks...done (149ms) 
     -computing association tests...done (0ms) 
     -computing joint association tests...done (0ms) 

Association results stored separately for each trait in files : 
* [[FILENAME]]

Number of ignored tests due to low MAC : 0

Masks written to : [[FILENAME].{bed,bim,fam}]

Elapsed time : 9.54679s
End time: Tue Apr  5 22:12:49 2022
`

**Comment by christopherdkoch** [comment]
This is all the gets printed in the standard error. Below is the standard output, though it doesn't look unusual to me. Yes, I ran ACATV also but did not get outputs for that either for this mask/gene. This is a QT phenotype. I can check if the problem exists for BTs.

              |=============================|
              |      REGENIE v3.0.1.gz      |
              |=============================|

Copyright (c) 2020-2022 Joelle Mbatchou, Andrey Ziyatdinov and Jonathan Marchini.
Distributed under the MIT License.
Compiled with Boost Iostream library.
Using Intel MKL with Eigen.

Log of output saved in file : [FILENAME]

Options in effect:
  --step 2 \
  --threads 1 \
  --bed [FILENAME] \
  --covarFile [FILENAME] \
  --phenoFile [FILENAME] \
  --phenoCol [PHENO] \
  --bsize 100 \
  --pred [FILENAME] \
  --out [FILENAME] \
  --aaf-bins 0.05 \
  --anno-file [FILENAME] \
  --set-list [FILENAME] \
  --mask-def [FILENAME] \
  --write-mask \
  --apply-rint \
  --minMAC 1 \
  --vc-tests skat,skato,acatv,acato \
  --vc-maxAAF 0.05 \
  --joint acat \
  --debug \
  --range 8:19759228-19824769

Association testing mode (joint tests) with fast multithreading using OpenMP
 * bim              : [[FILENAME]] n_snps = 4982410
   -number of variants after filtering on range = 300
 * fam              : [[FILENAME]] n_samples = 30108
 * bed              : [[FILENAME]]
 * phenotypes       : [[FILENAME]] n_pheno = 1
   -dropping observations with missing values at any of the phenotypes
   -number of phenotyped individuals with no missing data = 27519
 * covariates       : [[FILENAME]] n_cov = 14
   -number of individuals with covariate data = 30106
 * number of individuals used in analysis = 27519
   -applying RINT to all phenotypes
 * LOCO predictions : [[FILENAME]]
   -file [[FILENAME]] for phenotype '[PHENO]'
   -residualizing and scaling phenotypes...done (0ms) 
 * annotations      : [[FILENAME]] 
   +number of annotations categories = 7
 * masks            : [[FILENAME]] n_masks = 1
 * aaf cutoffs      : [ 1 : 0.05 ] + singletons
 * set file         : [[FILENAME]] n_sets = 1
WARNING: Detected 19432 sets with only unknown variants (these are ignored).
 * # threads        : [1]
 * # tested sets    : [1]
 * max block size   : [100]
 * rule used to build masks : max
 * computing gene-based tests for each set of variants included in a mask (rho=[0,0.01,0.04,0.09,0.16,0.25,0.5,1])
  -variants with MAC <= 10 are collapsed into a mask
  -weights are obtained from Beta(MAF,1,25)
 * approximate memory usage : 28MB
 * user specified to write masks (in PLINK bed format)
 * using minimum MAC of 1 (masks with lower MAC are ignored)

Chromosome 8 [1 sets in total]
   -reading loco predictions for the chromosome...done (19ms) 
 set [1/1] : 9 variants...1 chunks
     -reading in genotypes, computing gene-based tests and building masks...done (149ms) 
     -computing association tests...done (0ms) 
     -computing joint association tests...done (0ms) 

Association results stored separately for each trait in files : 
* [[FILENAME]]

Number of ignored tests due to low MAC : 0

Masks written to : [[FILENAME].{bed,bim,fam}]

Elapsed time : 9.54679s
End time: Tue Apr  5 22:12:49 2022

**Comment by joellembatchou** [comment]
Hi,

The output indicates that no variants are included in the variant sets for the SKAT/ACAT tests. Can you include the sum stats output in `[[FILENAME]` (as you mentioned burden masks results were printed)?

**Comment by christopherdkoch** [comment]
Hi Joelle,

Is this what you're looking for? The SKAT/ACAT tests are for mask1 at < 0.05, so I think should include the same variants as mask1.0.05 burden? I will also try with the most recent release.

CHROM GENPOS ID ALLELE0 ALLELE1 A1FREQ N TEST BETA SE CHISQ LOG10P EXTRA
8 19796972 LPL.mask1.singleton ref mask1.singleton 9.08463e-05 27519 ADD 0.130305 0.431942 0.0910069 0.117532 DF=1
8 19796972 LPL.mask1.0.05 ref mask1.0.05 0.000181693 27519 ADD 0.225722 0.30549 0.545952 0.337264 DF=1
8 19796972 LPL NA NA NA NA ADD-BURDEN-ACAT NA NA 0.295937 0.231776  DF=2

**Comment by christopherdkoch** [comment]
Does total AC need to be > 10 for SKAT/ACAT to run at all? Looks like AC=10 in this case. I can check other missing tests if that's a plausible explanation.

**Comment by joellembatchou** [comment]
Hi,

It should still run even if there are only variants with MAC<=10 (these will be collapsed into a mask which will be used as a pseudo-variant in the test. We have pushed a new update which addresses a couple bugs that were present for these tests ([v3.1](https://github.com/rgcgithub/regenie/releases/tag/v3.1)). Can you try with the new version (still using `--debug`)?

**Comment by christopherdkoch** [comment]
Hi Joelle,

Using v3.1 I'm getting the same output as before for my test cases, though there are a few extra lines in the log:

Q_SKAT for all masks:
0 0
Q_BURDEN for all masks:
0 0
#sites in mask=0
checking burden masks in set...# burden masks for joint acat test = 2
done building acat weights

**Comment by joellembatchou** [comment]
Can you try adding `--write-mask-snplist` to get the list of variants that go into the 5% AAF masks and then tests these as single variants in REGENIE specifying `--minMAC 1` (just to better understand what the AAF/p-values would be for those).

**Comment by christopherdkoch** [comment]
I've run Regenie with `--minMAC 1`, and below are the single variant results for those variants output by `--write-mask-snplist` (I removed the first 5 columns from the output). Also, I'm noticing there may have been a possible miscommunication earlier in the thread. Is it possible that the total (after collapsing) MAC needs to be greater than 10 for SKAT tests (I believe it's 10 in this example)?

```
1.83332e-05 27273 ADD -0.824811 0.965736 0.729445 0.405538 NA
5.45217e-05 27512 ADD -0.192482 0.557643 0.119142 0.136697 NA
3.64604e-05 27427 ADD 1.09085 0.683049 2.55052 0.95759 NA
1.82302e-05 27427 ADD -0.157267 0.965849 0.0265128 0.0601542 NA
1.82309e-05 27426 ADD 1.27277 0.965717 1.73701 0.726958 NA
1.81699e-05 27518 ADD 0.462793 0.965931 0.229552 0.199382 NA
1.82375e-05 27416 ADD -0.103335 0.965769 0.0114485 0.0386783 NA
```

**Comment by joellembatchou** [comment]
Hi,

This is very late but I think your issue should be fixed in [v3.1.4](https://github.com/rgcgithub/regenie/releases/tag/v3.1.4) (sets with only variants with MAC<10 were being skipped in SKAT/ACAT tests).

Cheers,
Joelle

**Comment by christopherdkoch** [comment]
Thanks for letting me know! I'll try it out soon and let you know if I run into any problems.
